rust/xain-coordinator/src/training.rs

```diff
@@ -0,0 +1,337 @@
+use std::{
+    any::Any,
+    collections::{BTreeMap, BTreeSet},
+    fmt,
+    time::Duration,
+};
+
+use crate::{Model, ModelDim};
+
+macro_rules! err {
+    ($($tt:tt)*) => {
+        ProtocolError { msg: format!($($tt)*) }
+    }
+}
+
+macro_rules! bail {
+    ($($tt:tt)*) => {
+        return Err(err!($($tt)*))
+    }
+}
+
+#[derive(Debug)]
+pub struct ProtocolError {
+    msg: String,
+}
+
+impl fmt::Display for ProtocolError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.msg, f)
+    }
+}
+
+impl std::error::Error for ProtocolError {}
+
+type ParticipantId = u32;
+
+#[derive(Debug)]
+pub struct FromParticipant<T> {
+    pub from: ParticipantId,
+    pub payload: T,
+}
+
+#[derive(Debug)]
+pub enum InMessage {
```

Comment:
Just out of interest: Do we also want to handle client timeouts through InMessage?

Reply:
Yeah, that's my plan. The coordinator will receive `InMessageKind::Leave` message if the client drops. It's the job of the transport layer to synthesize `Leave` if the client seems dead. That is, I think it makes sence to treat hearbeats as a transport layer concern.